### PR TITLE
Support backward for margin_ranking_loss and add tests

### LINKED ISSUE #2348: # Bug: margin_ranking_loss 缺少 Autograd 注册导致梯度链断裂
## 问题概述

**严重程度:** 高 (P1)
**问题类型:** FlagGems算子实现缺陷
**影响范围:** 所有依赖 margin_ranking_loss backward 的训练场景

## 问题描述

FlagGems 的 `margin_ranking_loss` 算子只注册了前向计算，没有注册对应的 autograd kernel。当 backward 传播经过此算子时，梯度链断裂，导致输入 tensor 的梯度为 None。

PyTorch 在 backward 时会发出以下警告：
```
aten::margin_ranking_loss: an autograd kernel was not registered to the Autograd key(s) 
but we are trying to backprop through it. This may lead to silently incorrect behavior.
This behavior is deprecated and will be removed in a future version of PyTorch.
```

## 复现步骤

```python
import torch
import flag_gems

flag_gems.enable()

anchor = torch.randn(4, device="cuda", requires_grad=True)
positive = torch.randn(4, device="cuda", requires_grad=True)
label = torch.ones(4, device="cuda")

loss_fn = torch.nn.MarginRankingLoss(margin=1.0)
loss = loss_fn(anchor, positive, label)
loss.backward()

print(anchor.grad)   # None — 预期应有梯度值
print(positive.grad)  # None — 预期应有梯度值
```

## 根因分析

**算子文件位置:** `src/flag_gems/ops/margin_ranking_loss.py`

**注册方式:** 在 `src/flag_gems/__init__.py:254` 通过 `("margin_ranking_loss", margin_ranking_loss)` 注册到 `aten::margin_ranking_loss`。

当前实现只提供了前向 kernel (`_margin_ranking_loss_kernel`)，通过 Triton 实现。但 `aten::margin_ranking_loss` 在 PyTorch 中是一个 composite 算子（本身由更基本的操作组合而成），原生 PyTorch 是通过 `CompositeImplicitAutograd` dispatch key 来自动获得 backward 支持的。

FlagGems 覆盖了这个算子但没有注册到 `Autograd` key，导致 autograd 路径被阻断。

## 是否需要 Fix

**需要** — 这是一个在实际训练场景中必然触发的缺陷：
- margin_ranking_loss 广泛用于排序学习、对比学习、度量学习等场景
- 这些场景必然需要 backward，否则无法训练
- 当前实现会**静默**地丢失梯度（只有 warning，不会报错），可能导致训练看似正常但模型实际不学习

## 建议修复方案

方案 1（推荐）：将 `margin_ranking_loss` 实现为 composite 算子，分解为 FlagGems 已经支持的基本操作 (sub, mul, clamp, mean/sum)，让 autograd 自动处理。

方案 2：实现自定义的 `torch.autograd.Function`，手动编写 forward 和 backward 的 Triton kernel。

## 单测覆盖分析

**现有单测:** `tests/test_special_ops.py:2466` — `test_accuracy_margin_ranking_loss`

**为什么没有发现该问题:**
- 单测只验证了前向精度（`gems_assert_close(res_out, ref_out, dtype)`），没有测试 backward
- 没有验证 `requires_grad=True` 输入通过算子后梯度是否正确传播
- 单测的设计模式是"前向精度对比"，不包含梯度正确性验证

**组合测例为什么能发现:**
- 组合测例将 `margin_ranking_loss` 放入多任务损失模型中
- 模型训练需要 `loss.backward()`，自然触发了 backward 路径
- 验证了所有 `requires_grad=True` 输入的 `.grad` 属性
- 这反映了组合测例相比单元测试更贴近实际使用场景的优势

## 影响的使用场景

| 场景 | 影响程度 | 说明 |
|------|---------|------|
| 排序学习 (Learning to Rank) | 高 | margin_ranking_loss 是核心损失函数 |
| 对比学习 | 中 | 常用于构建 triplet loss 等变体 |
| 推荐系统 | 中 | pairwise ranking 场景 |
| 度量学习 | 中 | margin-based loss 的基础 |

### PR DESCRIPTION
**Pull request by YY4994**
### PR Category
Operator | OP Test

### Type of Change
Bug Fix

### Description
Rewrite margin_ranking_loss as a composite of differentiable primitive ops so autograd can trace through the computation graph naturally. Add backward tests covering all reduction modes, shapes, and float dtypes.

### Issue

- Resolves #2348

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### CODE REVIEW

File: src/flag_gems/ops/margin_ranking_loss.py [hunk 1]
```diff
@@ -14,105 +14,296 @@
...
+
+        Args:
+            ctx: Context object with saved tensors from forward pass
+            grad_output: Gradient from upstream layers
+
+        Returns:
+            Tuple of gradients (grad_x1, grad_x2, None, None, None)
+            where None corresponds to target, margin, and reduction (no gradients needed)
+        """
+
```
Comment (tengqm) [inline src/flag_gems/ops/margin_ranking_loss.py:229]: Please add a logger call as we did in line 139.
This helps us make sure that we know it when the backward path is invoked.
`GEMS MARGIN_RANKING_LOSS_BACKWARD`.

File: src/flag_gems/ops/margin_ranking_loss.py [hunk 2]
```diff
@@ -14,105 +14,296 @@
...
+        original_shape = ctx.original_shape
+        grad_x1 = grad_x1.view(original_shape)
+        grad_x2 = grad_x2.view(original_shape)
+
+        # Return gradients (zero grad for target to support autograd.grad with allow_unused=False)
+        grad_target = torch.zeros_like(y).view(original_shape)
+        return grad_x1, grad_x2, grad_target, None, None
+
+
+def margin_ranking_loss(x1, x2, target, margin=0.0, reduction="mean"):
```
Comment (tengqm) [inline src/flag_gems/ops/margin_ranking_loss.py:281]: Based on https://github.com/flagos-ai/FlagGems/blob/master/src/flag_gems/__init__.py#L296

The forward and the backward cases together form a complete operator.
Neither case is supposed to be used as a standalone operator.
Am I understanding this correctly?

File: tests/test_margin_ranking_loss.py [hunk 3]
```diff
@@ -31,3 +31,72 @@ def test_margin_ranking_loss(shape, dtype, margin, reduction):
...
+
+REDUCTION_MAP = {0: "none", 1: "mean", 2: "sum"}
+
+
+@pytest.mark.margin_ranking_loss
+@pytest.mark.parametrize("shape", [(2, 3), (128, 256), (1024, 256)])
+@pytest.mark.parametrize("dtype", utils.FLOAT_DTYPES)
+@pytest.mark.parametrize("margin", [0.0, 0.5, 1.0])
+@pytest.mark.parametrize("reduction", [0, 1, 2])
+def test_margin_ranking_loss_backward(shape, dtype, margin, reduction):
```
Comment (tengqm) [inline tests/test_margin_ranking_loss.py:44]: I think we need a benchmark for the backward case as well.
Please double check.
Thanks.

File: benchmark/test_margin_ranking_loss.py [hunk 4]
```diff
@@ -22,3 +31,15 @@ def test_margin_ranking_loss():
         dtypes=consts.FLOAT_DTYPES,
     )
     bench.run()
+
+
+@pytest.mark.margin_ranking_loss
```
Comment (tengqm) [inline benchmark/test_margin_ranking_loss.py]:
```suggestion
@pytest.mark.margin_ranking_loss_backward
```

File: src/flag_gems/ops/margin_ranking_loss.py [hunk 5]
```diff
@@ -14,105 +14,297 @@
...
 
+    # Load input tensors with boundary checking
     x1 = tl.load(x1_ptr + offsets, mask=mask, other=0)
     x2 = tl.load(x2_ptr + offsets, mask=mask, other=0)
     y = tl.load(target_ptr + offsets, mask=mask, other=0)
 
+    # Compute margin ranking loss: max(0, -y * (x1 - x2) + margin)
     diff = x1 - x2
-    m = tl.full([BLOCK_SIZE], margin, x1.dtype)
+    m = tl.full([BLOCK_SIZE], margin, tl.float32)
```
Comment (tengqm) [inline src/flag_gems/ops/margin_ranking_loss.py:45]: Not sure if it is okay to hardcode the dtype here.

**Review by tengqm — COMMENTED** [review]
code style and completeness wise, I'm okay with this PR.

**Review by huangyiqun — APPROVED** [review]
